internal/client.go

```diff
@@ -470,64 +532,97 @@ const (
-	var dataConverter DataConverter
-	if options != nil && options.DataConverter != nil {
-		dataConverter = options.DataConverter
-	} else {
-		dataConverter = getDefaultDataConverter()
+
+	connection, err := options.GRPCDialer(GRPCDialerParams{
+		HostPort:             options.HostPort,
+		RequiredInterceptors: requiredInterceptors(metricsScope),
+		DefaultServiceConfig: DefaultServiceConfig,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return newServiceClient(workflowservice.NewWorkflowServiceClient(connection), connection, domain, options), nil
+}
+
+func newServiceClient(workflowServiceClient workflowservice.WorkflowServiceClient, connectionCloser io.Closer, domain string, options ClientOptions) Client {
+	options.MetricsScope = tagScope(options.MetricsScope, tagDomain, domain, clientImplHeaderName, clientImplHeaderValue)
+
+	if len(options.Identity) == 0 {
+		options.Identity = getWorkerIdentity("")
 	}
-	var contextPropagators []ContextPropagator
-	if options != nil {
-		contextPropagators = options.ContextPropagators
+
+	if options.DataConverter == nil {
+		options.DataConverter = getDefaultDataConverter()
 	}
-	var tracer opentracing.Tracer
-	if options != nil && options.Tracer != nil {
-		tracer = options.Tracer
-		contextPropagators = append(contextPropagators, NewTracingContextPropagator(zap.NewNop(), tracer))
+
+	if options.Tracer != nil {
+		options.ContextPropagators = append(options.ContextPropagators, NewTracingContextPropagator(zap.NewNop(), options.Tracer))
 	} else {
-		tracer = opentracing.NoopTracer{}
+		options.Tracer = opentracing.NoopTracer{}
 	}
+
 	return &workflowClient{
-		workflowService:    metrics.NewWorkflowServiceWrapper(rpc.NewWorkflowServiceErrorWrapper(service), metricScope),
+		workflowService:    workflowServiceClient,
+		connectionCloser:   connectionCloser,
 		domain:             domain,
 		registry:           newRegistry(),
-		metricsScope:       metrics.NewTaggedScope(metricScope),
-		identity:           identity,
-		dataConverter:      dataConverter,
-		contextPropagators: contextPropagators,
-		tracer:             tracer,
+		metricsScope:       metrics.NewTaggedScope(options.MetricsScope),
+		identity:           options.Identity,
+		dataConverter:      options.DataConverter,
+		contextPropagators: options.ContextPropagators,
+		tracer:             options.Tracer,
 	}
 }
 
 // NewDomainClient creates an instance of a domain client, to manager lifecycle of domains.
-func NewDomainClient(service workflowservice.WorkflowServiceClient, options *ClientOptions) DomainClient {
-	var identity string
-	if options == nil || options.Identity == "" {
-		identity = getWorkerIdentity("")
-	} else {
-		identity = options.Identity
+func NewDomainClient(options ClientOptions) (DomainClient, error) {
+	metricsScope := tagScope(options.MetricsScope, tagDomain, "domain-client", clientImplHeaderName, clientImplHeaderValue)
```

Comment:
I don't think using 'tagDomain' makes sense here.  You should remove this tag.
Or use some constant value like 'EmptyDomain'

Reply:
Remove.